### PR TITLE
Feature/Popover Definitions

### LINKED ISSUE #4117: Add definition popover feature to challenges
For specific terms, we can create bootstrap popovers to allow you view their definitions (rather than having to look them up).

Here's an example:

Here's the working code: http://jsfiddle.net/v5L7m/263/

The catch is we need to create JavaScript in main.js that will allow for an infinite number of links with popovers within the same challenge description.

### PR DESCRIPTION
**Pull request by Bouncey**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of FreeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](https://github.com/FreeCodeCamp/FreeCodeCamp/wiki/git-rebase#squashing-multiple-commits-into-one) them into one commit).
- [x] All new and existing tests pass the command `npm run test-challenges`. Use `git commit --amend` to amend any fixes.
#### Type of Change

- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)
#### Checklist:

- [x] Tested changes locally.
- [x] Closes currently open issue (replace XXXX with an issue no): Closes #4117
#### Description

Added popover definitions to one challenge only (`Say Hello to HTML Elements`) for review/feedback.

### CODE REVIEW

**Comment by Bouncey** [comment]
One concern I have is doing it this way makes the code look messy and harder to understand. I have a fair idea how to smarten it up with angular (custom directive), but not react.

**Comment by Bouncey** [comment]
Reworking, smartening up the look of the popover and the code needed. 

Please block until amended @erictleung @raisedadead

**Comment by erictleung** [comment]
@Bouncey thanks for the update. Some feedback on your changes, your solution so far (not sure if you were plan on changing it) doesn't appear to be scalable because the definition is within the challenge itself. What was suggested in that issue discussion was having a separate JSON file with all the definitions. Let us know if you have any questions.

**Comment by Bouncey** [comment]
Just updated with a different version. I removed most of what I could from the challenge seed, but the popover needs to come from either a `<button>` tag or `<a>` tag.

I aim to split the specific CSS into a different file as well for the final commit.

`def.js` was the easiest way I could see of populating the popover with the required content, and still have some readability in the challenge seed.

**Comment by Bouncey** [comment]
Please QA

**Comment by bayuanderas** [comment]
http://whow.simdif.com
Pada tanggal 10 Jun 2016 20.01, "CamperBot" <notifications@github.com>
menulis:

**Comment by erictleung** [comment]
@Bouncey there appears to be some merge conflicts now. I would rebase this branch onto an updated repo. If you've gotten your remotes set up correctly, this code should help you.

```bash
git checkout staging
git pull --rebase upstream staging
git checkout Feature/Popover-Def-HTML-CSS
git rebase staging
```

From there, you should be able to find the conflicts and make changes accordingly.

cc/ @FreeCodeCamp/issue-moderators any feedback on how @Bouncey is implementing the definitions?

**Comment by Bouncey** [comment]
@erictleung @raisedadead 

I have 
`git commit --amend`
`git push -f`

but TravisCI hasn't fired?

**Comment by Bouncey** [comment]
OK, please QA

**Comment by atjonathan** [comment]
@FreeCodeCamp/issue-moderators ^

I love the way it looks, you have a few spacing issue you could fix.
Maybe rename the file to `definitions-showHTML.js`

**Comment by raisedadead** [comment]
@atjonathan Looks like this needs re-basing.

File: client/less/main.less [hunk 1]
```diff
@@ -1161,6 +1161,41 @@ code {
     background-color: @brand-primary;
 }
 
+// for bootstrap popover
+.pop-explain code{
```
Comment (BerkeleyTrue) [inline client/less/main.less]: These new css class definitions should be in their own file

File: public/js/definitions-html5-css.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
+$(document).ready(function(){
```
Comment (BerkeleyTrue) [inline public/js/definitions-html5-css.js]: This file should be in client/commonFramework folder and follow the init pattern established there

File: server/views/challenges/showHTML.jade [hunk 3]
```diff
@@ -73,6 +73,9 @@ block content
                             else
                                 a#next-challenge.btn.btn-lg.btn-primary.btn-block(href="/challenges/next-challenge?id="+id) Go to my next challenge (ctrl + enter)
     include ../partials/challenge-modals
+    
+    script(src='/js/definitions-html5-css.js')
```
Comment (BerkeleyTrue) [inline server/views/challenges/showHTML.jade]: This should be combined into the common framework and not a separate file

File: client/less/main.less (hunk 1 above)
Comment (Bouncey) [inline client/less/main.less]: :+1:  Done

File: public/js/definitions-html5-css.js (hunk 2 above)
Comment (Bouncey) [inline public/js/definitions-html5-css.js]: I think I'm starting to get it, would I need to start the file with something like

```
window.common = (function(global) {
  const {
    $,
    Rx: { Observable },
    common = { init: [] },
    Mousetrap
  } = global;
```
?

Comment (Bouncey) [inline public/js/definitions-html5-css.js]: :+1: Got it

File: server/views/challenges/showHTML.jade (hunk 3 above)
Comment (Bouncey) [inline server/views/challenges/showHTML.jade]: :+1:

**Comment by Bouncey** [comment]
Just re-basing, and will amend.

File: seed/challenges/01-front-end-development-certification/basic-javascript.json [hunk 4]
```diff
@@ -3385,7 +3385,7 @@
       "id": "5664820f61c48e80c9fa476c",
       "title": "Golf Code",
       "description": [
-        "In the game of <a href=\"https://en.wikipedia.org/wiki/Golf\" target=\"_blank\">golf</a> each hole has a <code>par</code> for the average number of <code>strokes</code> needed to sink the ball. Depending on how far above or below <code>par</code> your <code>strokes</code> are, there is a different nickname.",
+        "In the game of <a href=\"https://en.wikipedia.org/wiki/Golf\" target=\"_blank\">golf</a> each hole has a <code>par</code> meaning the average number of <code>strokes</code> a golfer is expected to make in order to sink the ball in a hole to complete the play. Depending on how far above or below <code>par</code> your <code>strokes</code> are, there is a different nickname.",
```
Comment (Bouncey) [inline seed/challenges/01-front-end-development-certification/basic-javascript.json]: I have tried re-basing and reset hard, but cannot remove this file from my commit

**Comment by Bouncey** [comment]
Oh, I now have 24 commits..... I did something wrong...

File: gulpfile.js [hunk 5]
```diff
@@ -132,7 +132,8 @@ var paths = {
     'run-tests-stream',
     'show-completion',
     'step-challenge',
-    'end'
+    'end',
+    'definitions-html5-css'
```
Comment (BerkeleyTrue) [inline gulpfile.js]: This should come before 'end'

**Comment by BerkeleyTrue** [comment]
Looks like you brought in a bunch of commits. You may want to start over with a fresh branch

**Comment by Bouncey** [comment]
I fixed it now

**Comment by Bouncey** [comment]
Ok, Ill amend the gulpfile

File: client/commonFramework/definitions-html5-css.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
+window.common = (function(global) {
+const {
+$,
+common = { init: [] }
+} = global;
+
+// Say Hello to HMTL Elements
```
Comment (atjonathan) [inline client/commonFramework/definitions-html5-css.js]: ~HMTL Elements~~ HTML Elements

File: client/less/popover.less [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+  border-radius: 6px 6px 0 0;
+  left: 15% !important;
+}
+.popover-title {
+    margin: 0;
+    padding: 8px 14px;
+    font-size: inherit;
+    background-color: @brand-primary;
+    border-radius: 5px 5px 0 0;
+    color: #fff;
```
Comment (atjonathan) [inline client/less/popover.less]: You are using tabs here, when you were using spaces earlier.

File: seed/challenges/01-front-end-development-certification/html5-and-css.json [hunk 8]
```diff
@@ -1415,7 +1416,7 @@
       ],
       "tests": [
         "assert((/cat photos/gi).test($(\"a\").text()), 'message: Your <code>a</code> element should have the <code>anchor text</code> of \"cat photos\".');",
-        "assert(/http:\\/\\/freecatphotoapp\\.com/gi.test($(\"a\").attr(\"href\")), 'message: You need an <code>a</code> element that links to <code>http&#58;//freecatphotoapp<wbr>.com</code>');",
+        "assert(/http:\\/\\/freecatphotoapp\\.com/gi.test($(\"a\").attr(\"href\")), 'message: You need an <code>a</code> element that links to <code>http&#58;//freecatphotoapp.com</code>');",
```
Comment (atjonathan) [inline seed/challenges/01-front-end-development-certification/html5-and-css.json]: @Bouncey, why was `<wbr>` removed? This was recently implemented, and shouldn't be removed.

File: seed/challenges/01-front-end-development-certification/html5-and-css.json [hunk 9]
```diff
@@ -4532,7 +4533,7 @@
         "</style>"
       ],
       "tests": [
-        "assert($(\"body\").css(\"background-color\") === \"rgb(0, 0, 0)\", 'message: Your <code>body</code> element should have a black background.');",
+        "assert($(\"body\").css(\"background-color\") === \"rgb(0, 0, 0)\", 'message: Your <code>body</code> element should have an black background.');",
```
Comment (atjonathan) [inline seed/challenges/01-front-end-development-certification/html5-and-css.json]: ~~an black background.~~ **a** black background.

File: seed/challenges/01-front-end-development-certification/html5-and-css.json (hunk 8 above)
Comment (Bouncey) [inline seed/challenges/01-front-end-development-certification/html5-and-css.json]: @atjonathan I didn't? I've only just rebased the branch too?

I'll add it back in

Comment (atjonathan) [inline seed/challenges/01-front-end-development-certification/html5-and-css.json]: @Bouncey, probably due to your `git` issues earlier.

File: seed/challenges/01-front-end-development-certification/html5-and-css.json (hunk 9 above)
Comment (Bouncey) [inline seed/challenges/01-front-end-development-certification/html5-and-css.json]: I think this is related to the <wbr> issue as well

Comment (atjonathan) [inline seed/challenges/01-front-end-development-certification/html5-and-css.json]: Yeh it is :smile:

**Comment by atjonathan** [comment]
@Bouncey, these look good! You could start working on the next bunch of challenges :+1:

**Comment by Bouncey** [comment]
I have read through every challenge in `HTML5 and CSS`, `Bootstrap` and `jquery'. I'm stuggling to find new concepts that aren't well explained in the challenge instructions.

**Comment by atjonathan** [comment]
@Bouncey, it's possible we could eliminate some of the instructions, and replace it with popovers.

**Comment by Bouncey** [comment]
@atjonathan You'll lose some of the context of the instructions, so you end up having to re-write the challenge in order to have a popover. A lot of effort...

**Comment by Bouncey** [comment]
I think the format is established enough for other new concepts to be caught with issues

**Comment by atjonathan** [comment]
@Bouncey, can this be merged or are you going to add more?

**Comment by BerkeleyTrue** [comment]
I'd like to take a look at this before it's merged

**Comment by Bouncey** [comment]
Any feedback @BerkeleyTrue ?

**Comment by BerkeleyTrue** [comment]
Sorry about that @Bouncey I haven't had the chance. My top priority is moving everything over to React.

Do you mind adding some gifs of what it looks like?

@atjonathan Did you QA this?

**Comment by Bouncey** [comment]
@BerkeleyTrue I understand you not getting round to it, not a problem.

anim

**Comment by Bouncey** [comment]
@BerkeleyTrue And mobile

anim

**Comment by atjonathan** [comment]
I will `QA` this later today @BerkeleyTrue.

**Comment by atjonathan** [comment]
### **LGTM :+1:**
Great job on this @Bouncey :tada: :stuck_out_tongue_winking_eye: 
@QuincyLarson, please take a look before this is merged...

**Comment by atjonathan** [comment]
Alright, merging as looks good and the user interface can always be changed later :+1:

**Comment by BerkeleyTrue** [comment]
@atjonathan This was waiting on @QuincyLarson to comment

**Comment by BerkeleyTrue** [comment]
I've reverted the changes. @Bouncey please feel free to reopen the PR and wait for @QuincyLarson's comment.

Happy Coding

**Comment by atjonathan** [comment]
@BerkeleyTrue, Np I was merging after a while of inactivity and multiple issue mods had tested.

**Pull request merged**